lib/openstax/book.rb

```diff
@@ -0,0 +1,83 @@
+require_relative './page'
+
+module Openstax
+  # A Openstax::Book is the object that "represents" the book to the search
+  # world.  A book is composed of pages, and pages are composed of elements.
+  #
+  # An indexing stategy is injected in order to determine what elements we
+  # parse out of the page.
+  #
+  # Pages and Elements are lazy loaded.
+  class Book
+    CNX_URL_PREFIX = 'https://archive.cnx.org/contents'
+
+    attr_reader :uuid, :version
+
+    def initialize(uuid:, version:, indexing_strategy:)
+      @uuid = uuid
+      @version = version
+      @indexing_strategy = indexing_strategy
+    end
+
+    def pages
+      @pages ||= begin
+        page_ids = recursive_flatten_page_ids(data['tree']['contents'])
+        Rails.logger.info "OpenSearch: found #{page_ids.count} possible page_ids in book #{@uuid}@#{@version}"
+
+        pages = page_ids.map do |page_id|
+          response = fetch_cnx_url(page_uri(page_id)) rescue next
```

Comment:
@jpslav we went thru a lot of this at my last job.  best ways to handle errors like this.  imo, retries usually is a waste of time and effort and messing the code up (if something fails, it's likely it'll fail again).   retries have a place (e.g., in the financial world of transactions) but here...my 2 cent is that we should record the exact information we need, report it to honeybadger (or whatever we use), and move on.  missing out on a page is not the end of the world, better to be sure to get the rest.  And have that information handy for us in excception reporting so that we can see number of times it's happened and relevant page information so we can debug/fix.